spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraStrategies.scala

```diff
@@ -24,150 +23,15 @@ private[cassandra] trait CassandraStrategies {
-       * 4. Consecutive clustering columns must be used, but, contrary to partition key, the tail can be skipped.
-       */
-      private val clusteringColumnPredicatesToPushDown: Seq[Expression] = {
-        val (eqColumns, otherColumns) = clusteringColumns.span(eqPredicatesByName.contains)
-        val eqPredicates = eqColumns.flatMap(eqPredicatesByName)
-        val optionalNonEqPredicate = for {
-          c <- otherColumns.headOption.toSeq
-          p <- firstNonEmptySeq(rangePredicatesByName(c), inPredicatesByName(c).filter(
-            _ => c==clusteringColumns.last))
-        } yield p
-
-        eqPredicates ++ optionalNonEqPredicate
-      }
-
-      /**
-       * Selects indexed and regular column predicates for pushdown:
-       * 1. At least one indexed column must be present in an equality predicate to be pushed down.
-       * 2. Regular column predicates can be either equality or range predicates.
-       * 3. If multiple predicates use the same column, equality predicates are preferred over range predicates.
-       */
-      private val indexedColumnPredicatesToPushDown: Seq[Expression] = {
-        val inPredicateInPrimaryKey = partitionKeyPredicatesToPushDown.exists(isIn)
-        val eqIndexedColumns = indexedColumns.filter(eqPredicatesByName.contains)
-        val eqIndexedPredicates = eqIndexedColumns.flatMap(eqPredicatesByName)
-        // Don't include partition predicates as None-indexed predicates if partition predicates can't
-        // be pushed down because we use token range query which already has partition columns in the
-        // where clause and it can't have other partial partition columns in where clause any more.
-        val nonIndexedPredicates = for {
-          c <- allColumns if !partitionKeyPredicatesToPushDown.isEmpty && !eqIndexedColumns.contains(c) ||
-                              partitionKeyPredicatesToPushDown.isEmpty && !eqIndexedColumns.contains(c) &&
-                               !partitionKeyColumns.contains(c)
-          p <- firstNonEmptySeq(eqPredicatesByName(c), rangePredicatesByName(c))
-        } yield p
-
-        if (!inPredicateInPrimaryKey && eqIndexedColumns.nonEmpty)
-          eqIndexedPredicates ++ nonIndexedPredicates
-        else
-          Nil
-      }
-
-      val predicatesToPushDown = (
-        partitionKeyPredicatesToPushDown ++
-        clusteringColumnPredicatesToPushDown ++
-        indexedColumnPredicatesToPushDown).distinct
-
-      val predicatesToPreserve = predicates.filterNot(predicatesToPushDown.toSet.contains)
-    }
-
-
     def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
       case PhysicalOperation(projectList, predicates, relation: CassandraRelation) =>
         logInfo(s"projectList: ${projectList.toString()}")
         logInfo(s"predicates: ${predicates.toString()}")
-        val pushDown = new PredicatePushDown(predicates, relation)
+        val pushDown = new PredicatePushDown(predicates, relation.tableDef)
```

Comment:
Have you though about decoupling things even further by a `PredicatePushDownBuilder`, that is going to be injected into this component and used for constructing instance of `PredicatePushDown`?  This would allow us to extend testability of `CassandraStrategies`.